sync15-adapter/src/state.rs

```diff
@@ -0,0 +1,48 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::collections::HashMap;
+
+use bso_record::BsoRecord;
+use collection_keys::CollectionKeys;
+use error::{self, ErrorKind};
+use key_bundle::KeyBundle;
+use record_types::MetaGlobalRecord;
+use request::InfoConfiguration;
+use util::{ServerTimestamp, SERVER_EPOCH};
+
+/// Holds global Sync state, including server upload limits, and the
+/// last-fetched collection modified times, `meta/global` record, and
+/// collection encryption keys.
+#[derive(Debug)]
+pub struct GlobalState {
+    pub config: InfoConfiguration,
+    pub collections: HashMap<String, ServerTimestamp>,
+    pub global: Option<BsoRecord<MetaGlobalRecord>>,
```

Comment:
It does implement it. The as_ref calls are required since otherwise Option::map takes ownership of the inner value, which is not what we want. The noise is unfortunate but there's no way around it.